### DISCUSSION TITLE
is possible make click throught window?

### DISCUSSION THREAD

**Discussion by jdemaria22** [hajimehoshi/ebiten#2059] [Q&A]
Hi guys, i'm trying to create a overlay with ebiten.

I use SetScreenTransparent() /SetWindowDecorated(false), to create a window transparent a undecorated.

I use SetWindowFloating(false) to hold up the window all the time.

I'ts posible to make click throught the window?  

for example in C

https://stackoverflow.com/questions/39855720/windows-forms-pass-clicks-through-a-partially-transparent-always-on-top-window
Thanks

**Comment by hajimehoshi** [comment]
Unfortunately no so far. You can calculate this by `CurrentPosition` and `WindowPosition` and `DeviceScaleFactor` by yourself.

A problem of exposing the position though a window is that coordinate systems differ among OSes and it is tricky to offer an consistent way. (e.g. is the position affected by a zoom scale for Hi-DPI?)

**Reply by jdemaria22** [reply]
Thanks for the answering.
Just i need to draw circles or rect/lines or images on the main screen all the time. 

I need to move between the different windows and be able to see the draws.

its possible use SetWindowLong? to change attributes of the Windows. 

thanks!!

**Reply by hajimehoshi** [reply]
Yes, I think you can get an window handler from `GetActiveWindow` and do what you want :-) Calling Win32API from Go is a little tricky but possible. See https://github.com/hajimehoshi/ebiten/blob/main/internal/ui/ui_glfw_windows.go or https://pkg.go.dev/golang.org/x/sys/windows as references.

Thanks,

**Reply by jdemaria22** [reply]
Thanks for answering.

Yes i use win32api for read procces memory and other things.

I already tried to use SetWindowLong in the ebiten window, but the window crash and i need to finish de process with the taskbar.

I call SetWindowLong after calling ebiten.RunGame(), its correct?

**Reply by hajimehoshi** [reply]
> I call SetWindowLong after calling ebiten.RunGame(), its correct?

Right, or any correct window should not exist.

**Reply by jdemaria22** [reply]
Hi!, thanks for the answering.

In the func Update(), i call SetWindowLong() ones
and the program crash

mi func Update

func (g *Game) Update() error {
	if ebiten.IsKeyPressed(ebiten.KeyQ) {
		if count == 0 {
			yogur := win.FindWindow(nil, hook.StringToUTF16PtrElseNil("yogur"))
			style := win.GetWindowLong(yogur, win.GWL_STYLE)
			win.SetWindowLong(yogur, win.GWL_STYLE, style|win.WS_EX_LAYERED|win.WS_EX_TRANSPARENT)
			fmt.Println("yogur , ", win.GetWindowLong(yogur, win.GWL_STYLE))
			count++
		}

	}
	return nil
}

mi func main()

func main() {

	ebiten.SetWindowSize(screenWidth, screenHeight)
	ebiten.SetWindowTitle("yogur")
	ebiten.SetWindowResizable(false)
	ebiten.SetWindowDecorated(false)
	ebiten.SetScreenTransparent(true)
	ebiten.SetWindowFloating(false)
	err = ebiten.RunGame(NewGame())
	if err != nil {
		log.Fatal(err)
	}
}


The program start okey, but when "Q" is pressed just the game crashing
	
![image](https://user-images.githubusercontent.com/43354137/162224194-1c3a4b7e-603f-4357-a8da-9abf0989bc5c.png)

**Reply by hajimehoshi** [reply]
What is the crash message?

**Reply by hajimehoshi** [reply]
I think you can get a crash message if you run the application with `go run`

**Reply by jdemaria22** [reply]
No, the application run ok.

the window open and work perfect.

Sorry, when i call SetWindowLong, the application doesn't respond anymore.

And i need to finish the process

**Reply by jdemaria22** [reply]
I dont get crash, sorry

**Reply by jdemaria22** [reply]
Example, 

func (g *Game) Update() error {
	fmt.Println("ok...")
	if ebiten.IsKeyPressed(ebiten.KeyQ) {
		if count == 0 {
			yogur := win.FindWindow(nil, hook.StringToUTF16PtrElseNil("yogur"))
			fmt.Println("call setWindowLong: ", yogur)
			style := win.GetWindowLongPtr(yogur, win.GWL_STYLE)
			win.SetWindowLongPtr(yogur, win.GWL_STYLE, style|win.WS_EX_LAYERED|win.WS_EX_TRANSPARENT)
			count++
		}
	}
	return nil
}

when i press Q, call the SetWindownLong 
application doesn't response animore

![image](https://user-images.githubusercontent.com/43354137/162231661-d5e205bd-208d-4160-b101-0d9a8bcb90a2.png)

**Reply by hajimehoshi** [reply]
Let's see SetWindowLongPtr's returning value and the error value.

**Reply by jdemaria22** [reply]
Okey, 

I tried with the function SetWindowLong from win32api,

func SetWindowLong(hWnd HWND, index, value int32) (int32, error) {
	ret, _, err := syscall.Syscall(setWindowLong.Addr(), 3,
		uintptr(hWnd),
		uintptr(index),
		uintptr(value))
	if ret == 0 {
		return int32(ret), err
	}
	return int32(ret), nil
}

and my functión update 

func (g *Game) Update() error {
	fmt.Println("ok...")
	if ebiten.IsKeyPressed(ebiten.KeyQ) {
		if count == 0 {
			yogur := win.FindWindow(nil, hook.StringToUTF16PtrElseNil("yogur"))
			fmt.Println("call setWindowLong: ", yogur)
			style := win.GetWindowLong(yogur, win.GWL_STYLE)
			r, err := win.SetWindowLong(yogur, win.GWL_STYLE, style|win.WS_EX_LAYERED|win.WS_EX_TRANSPARENT)
			if err != nil {
				fmt.Println("error in  setWindowLong: ", err, r)
			}
			fmt.Println("setWindowLong : ", r)
			count++
		}
	}
	return nil
}

and the same problem, and nathing logs affter the call the function. As if the process is waiting for the Syscall to return

**Reply by hajimehoshi** [reply]
What library are you using to call SetWindowLong?

Unfortunately I am not an expert of Win32API, so I might not be able to have a good answer. Sorry.

**Reply by jdemaria22** [reply]
I use https://github.com/lxn/win for win32api. 

Also, thank you very much for help.

I will investigate more and tell you.

Thanksss

**Reply by jdemaria22** [reply]
Hi,
Problem solved using SetWindownLong affter create the ebiten window

wnd := win.FindWindow(nil, hook.StringToUTF16PtrElseNil(name))
	r2, err := win.SetWindowLong(wnd, win.GWL_EXSTYLE, win.WS_EX_COMPOSITED|win.WS_EX_LAYERED|win.WS_EX_TRANSPARENT|win.WS_EX_TOOLWINDOW|win.WS_EX_TOPMOST)
	if err != nil {
		fmt.Println("error in  setWindowLong: ", err, r2)
	}
	
with this code, i make click throught the window

*name* -> is the name of the ebiten window.

Thanks!
